storage: skip extra encoding in pebbleMVCCScanner

aadityasondhi opened the pull request:
Previously, we were re-encoding decoded keys in `pebbleMVCCScanner`.

This unnecessary encoding showed up on CPU profiles.

This change avoids the extra encoding by using `rawKey` instead of
`MVCCKey`.

Fixes #49145

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54699)
<!-- Reviewable:end -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, and @jbowens)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 587 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MHvb1FRBteiehP3Kpuv:-MHvb1FRBteiehP3Kpuw:b-586bkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/24cd5fa1b6a04ed18a69cfd2d80b6bb77c763e41/pkg/storage/pebble_mvcc_scanner.go#L587)):*
> ```Go
> 	if len(val) > 0 || p.tombstones {
> 		type unsafeRawKeyGetter interface { unsafeRawKey() []byte }
> 		rawKey := p.parent.(unsafeRawKeyGetter).unsafeRawKey()
> ```

I'll be interested to see benchmarks on this. I know there is a bit of expense inside the Go runtime to type assert to a different interface. And then we're calling `unsafeRawKey` while we've already retrieved the key. Normally this would be fine, but this is performance sensitive code and we're looking to eek out every last iota of speed. I was imagining that this PR would mainly be a revert of the `pebble_mvcc_scanner.go` changes in https://reviewable.io/reviews/cockroachdb/cockroach/42210, and then figuring out how to provide `unsafeRawKey` on the `Iterator` interface.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 587 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MHvb1FRBteiehP3Kpuv:-MHwnDB43kr1LM3kvka6:b-n3fhre) ([raw file](https://github.com/cockroachdb/cockroach/blob/24cd5fa1b6a04ed18a69cfd2d80b6bb77c763e41/pkg/storage/pebble_mvcc_scanner.go#L587)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'll be interested to see benchmarks on this. I know there is a bit of expense inside the Go runtime to type assert to a different interface. And then we're calling `unsafeRawKey` while we've already retrieved the key. Normally this would be fine, but this is performance sensitive code and we're looking to eek out every last iota of speed. I was imagining that this PR would mainly be a revert of the `pebble_mvcc_scanner.go` changes in https://reviewable.io/reviews/cockroachdb/cockroach/42210, and then figuring out how to provide `unsafeRawKey` on the `Iterator` interface.
</blockquote></details>

This implementation was borrowed from how it is done in `pebble_batch.go` https://github.com/aadityasondhi/cockroach/blob/storage-skip-extra-MVCC-encoding/pkg/storage/pebble_batch.go#L273. 

I see how this could affect performance though. I will try to implement it as you described instead. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, and @jbowens)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 587 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MHvb1FRBteiehP3Kpuv:-MHwsn29EDN2C5qdvVyT:blvwxa1) ([raw file](https://github.com/cockroachdb/cockroach/blob/24cd5fa1b6a04ed18a69cfd2d80b6bb77c763e41/pkg/storage/pebble_mvcc_scanner.go#L587)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

This implementation was borrowed from how it is done in `pebble_batch.go` https://github.com/aadityasondhi/cockroach/blob/storage-skip-extra-MVCC-encoding/pkg/storage/pebble_batch.go#L273. 

I see how this could affect performance though. I will try to implement it as you described instead. 
</blockquote></details>

Got it. The implementation in `clearIterRange` is not performance sensitive. There it is fine to do what is convenient and ergonomic. The code here is extremely performance sensitive. We're trying to eek out a few percentage points of improvement. In such situations it is ok to have slightly awkward code that is higher performance.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
Great progress so far! Think this is mostly done

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-:-MI0Wus8E36qxm9IRlXb:b-gf0wn1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, and @jbowens)

___
*[pkg/storage/pebble_iterator.go, line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI0VUAHDavc4H7ppGWh:-MI0VUAHDavc4H7ppGWi:b-9c136o) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b6dabed3269c933c66bba722f80032faced0b1/pkg/storage/pebble_iterator.go#L247)):*
> ```Go
> // UnsafeRawKey returns the raw key.
> func (p *pebbleIterator) UnsafeRawKey() []byte {
> 	return p.unsafeRawKey()
> ```

Better idea to just rename `unsafeRawKey` to `UnsafeRawKey` instead of having two methods.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 735 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI0W5J75TQTnoiyt2Vm:-MI0W5J75TQTnoiyt2Vn:bdegtcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b6dabed3269c933c66bba722f80032faced0b1/pkg/storage/pebble_mvcc_scanner.go#L735)):*
> ```Go
> 		p.savedBuf = append(p.savedBuf[:0], p.curRawKey...)
> 		p.savedBuf = append(p.savedBuf, p.curValue...)
> 		p.curKey, _ = decodeMVCCKey(p.savedBuf[:len(p.curRawKey)])
> ```

You'd also have to set `p.curRawKey = p.savedBuf[:len(p.curRawKey)]`, and then this call just becomes `p.curKey = decodeMVCCKey(p.curRawKey)`. Right now you're copying the raw key but not changing the reference after the copy, so `p.curRawKey` still points into the unsafe key in the iterator. This is probably why you got the test failure.

We could also potentially take advantage of the fact that `p.curKey.Key` is always a prefix of `p.curRawKey`, and get a tiny performance boost by doing `p.curKey.Key = p.curRawKey[:len(p.curKey.Key)]` instead of the decode call.  Might be worth seeing if this makes a difference in benchmarks.
___
*[pkg/storage/tee.go, line 1178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI0WkCWCoUFyPj-Mrs_:-MI0WkCWCoUFyPj-Mrsa:b-3i8n87) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b6dabed3269c933c66bba722f80032faced0b1/pkg/storage/tee.go#L1178)):*
> ```Go
> // UnsafeRawKey is not implemented for tee
> func (t *TeeEngineIter) UnsafeRawKey() []byte {
> 	return nil
> ```

Also merge this with `unsafeRawKey`. 


<!-- Sent from Reviewable.io -->

Commit:
storage: skip extra encoding in pebbleMVCCScanner

Previously, we were re-encoding decoded keys in `pebbleMVCCScanner`.

This unnecessary encoding showed up on CPU profiles.

This change avoids the extra encoding by using `rawKey` instead of
MVCCKey.

Fixes #49145

Release note: None

aadityasondhi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, and @jbowens)

___
*[pkg/storage/pebble_iterator.go, line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI0VUAHDavc4H7ppGWh:-MI126QQBpukpoEti1MT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b6dabed3269c933c66bba722f80032faced0b1/pkg/storage/pebble_iterator.go#L247)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Better idea to just rename `unsafeRawKey` to `UnsafeRawKey` instead of having two methods.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 735 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI0W5J75TQTnoiyt2Vm:-MI11w1IB6W0cdmmh5nT:b-pm0kws) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b6dabed3269c933c66bba722f80032faced0b1/pkg/storage/pebble_mvcc_scanner.go#L735)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

You'd also have to set `p.curRawKey = p.savedBuf[:len(p.curRawKey)]`, and then this call just becomes `p.curKey = decodeMVCCKey(p.curRawKey)`. Right now you're copying the raw key but not changing the reference after the copy, so `p.curRawKey` still points into the unsafe key in the iterator. This is probably why you got the test failure.

We could also potentially take advantage of the fact that `p.curKey.Key` is always a prefix of `p.curRawKey`, and get a tiny performance boost by doing `p.curKey.Key = p.curRawKey[:len(p.curKey.Key)]` instead of the decode call.  Might be worth seeing if this makes a difference in benchmarks.
</blockquote></details>

Thanks, will try to compare the results once I can get these tests to pass. 
___
*[pkg/storage/tee.go, line 1178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI0WkCWCoUFyPj-Mrs_:-MI11tdtDC5SNH3HnA-a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b6dabed3269c933c66bba722f80032faced0b1/pkg/storage/tee.go#L1178)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Also merge this with `unsafeRawKey`. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
This generally looks like what I expected to see, though I have a request. The code in `pebble_mvcc_scanner.go` is particularly subtle and performance sensitive. It is worth going through this diff and and the changes in https://reviewable.io/reviews/cockroachdb/cockroach/42210 line-by-line and calling out any differences in this PR to aid the reviewer. This is not a typical ask in a code review, but this is not typical code.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-:-MI1yLL64q20g9P7vcpy:b-bh5ro4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, and @jbowens)

___
*[pkg/storage/engine.go, line 95 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI1xBEqBgtVkBTdSlpB:-MI1xBEqBgtVkBTdSlpC:b6gasfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/engine.go#L95)):*
> ```Go
> 	// Key returns the current key.
> 	Key() MVCCKey
> 	// UnsafeRawKey returns the current raw key.
> ```

It is worth explaining what a raw key is here. Something like: "returns the current raw key (a.k.a. the encoded MVCC key)".


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, and @jbowens)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 386 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI4em9F-ZFRlKRGRC4B:-MI4em9G9RgvtxObPem9:bofcy62) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/pebble_mvcc_scanner.go#L386)):*
> ```Go
> 		}
> 		p.keyBuf = EncodeKeyToBuf(p.keyBuf[:0], p.curKey)
> 		p.err = p.intents.Set(p.keyBuf, p.curValue, nil)
> ```

You can skip this encoding and reuse `p.curRawKey` here.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 407 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI4eicC645RTSvoPfGc:-MI4eicD3gKTpjWPJrX6:bofcy62) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/pebble_mvcc_scanner.go#L407)):*
> ```Go
> 		// in the scan range.
> 		p.keyBuf = EncodeKeyToBuf(p.keyBuf[:0], p.curKey)
> 		p.err = p.intents.Set(p.keyBuf, p.curValue, nil)
> ```

You can skip this encoding and reuse `p.curRawKey` here.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 736 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI4eK-_AO6Tx6n-l2t8:-MI4eK-_AO6Tx6n-l2t9:brl9aky) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/pebble_mvcc_scanner.go#L736)):*
> ```Go
> 		p.savedBuf = append(p.savedBuf, p.curValue...)
> 		p.curRawKey = p.savedBuf[:len(p.curRawKey)]
> 		p.curKey, _ = decodeMVCCKey(p.curRawKey)
> ```

I realized that since you aren't really concerned about the timestamp, you can replace this with `p.curKey.Key = enginepb.SplitMVCCKey(p.curRawKey)`. decodeMVCCKey also calls that method internally, and on top of that it does timestamp decoding, which is unnecessary here as the timestamp is stored as a primitive value and is already copied.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 737 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI4eZukBa2nlzyxnnaJ:-MI4eZukBa2nlzyxnnaK:b-tdocn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/pebble_mvcc_scanner.go#L737)):*
> ```Go
> 		p.curRawKey = p.savedBuf[:len(p.curRawKey)]
> 		p.curKey, _ = decodeMVCCKey(p.curRawKey)
> 		p.curValue = p.savedBuf[len(p.curKey.Key):]
> ```

This would change to be `p.savedBuf[len(p.curRawKey):])`, and is probably why tests are failing.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
@petermattis Thank you for the review. I am still troubleshooting some of the tests, but I will do the line-by-line review once I can get this to work. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-:-MIJvbU90pYUrMpQp7O5:bxu6vyj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, @jbowens, and @petermattis)

___
*[pkg/storage/engine.go, line 95 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI1xBEqBgtVkBTdSlpB:-MIJvtD499Tcc3E1Uo4m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/engine.go#L95)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

It is worth explaining what a raw key is here. Something like: "returns the current raw key (a.k.a. the encoded MVCC key)".
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 587 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MHvb1FRBteiehP3Kpuv:-MIJw10FE_mQTikhwBrD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/24cd5fa1b6a04ed18a69cfd2d80b6bb77c763e41/pkg/storage/pebble_mvcc_scanner.go#L587)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Got it. The implementation in `clearIterRange` is not performance sensitive. There it is fine to do what is convenient and ergonomic. The code here is extremely performance sensitive. We're trying to eek out a few percentage points of improvement. In such situations it is ok to have slightly awkward code that is higher performance.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 386 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI4em9F-ZFRlKRGRC4B:-MIJvva0CA1M79VKfqRM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/pebble_mvcc_scanner.go#L386)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

You can skip this encoding and reuse `p.curRawKey` here.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 736 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI4eK-_AO6Tx6n-l2t8:-MIJw3ysCMQ5LPLym8il:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/pebble_mvcc_scanner.go#L736)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I realized that since you aren't really concerned about the timestamp, you can replace this with `p.curKey.Key = enginepb.SplitMVCCKey(p.curRawKey)`. decodeMVCCKey also calls that method internally, and on top of that it does timestamp decoding, which is unnecessary here as the timestamp is stored as a primitive value and is already copied.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 737 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI4eZukBa2nlzyxnnaJ:-MIJw4NI60G9EuhFafq3:b-5uq734) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/pebble_mvcc_scanner.go#L737)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

This would change to be `p.savedBuf[len(p.curRawKey):])`, and is probably why tests are failing.
</blockquote></details>

Made the change, but tests were still failing. I am trying to debug but haven't had any luck yet.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
Maybe this set of fixes makes the tests pass. Let me know how it goes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-:-MIK0q1iBGe58_Yf3EvK:bm22u38)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, @jbowens, and @petermattis)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 385 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIK0C1Y7p9o0dWKkNtv:-MIK0C1Y7p9o0dWKkNtw:b-6awmbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3ab85c89003ea9ec13b0664e8b0c8d1c126cbb/pkg/storage/pebble_mvcc_scanner.go#L385)):*
> ```Go
> 			return false
> 		}
> 		p.keyBuf = p.curRawKey
> ```

`p.keyBuf` is meant to always point to a clean buffer owned by `pebbleMVCCScanner` and not by the iterator. So you'd want to remove this assignment and use `p.curRawKey` directly below.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 406 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIK0my_BUj2imaxN-7M:-MIK0my_BUj2imaxN-7N:b-8bi3es) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3ab85c89003ea9ec13b0664e8b0c8d1c126cbb/pkg/storage/pebble_mvcc_scanner.go#L406)):*
> ```Go
> 		// continue scanning so that we can return all of the intents
> 		// in the scan range.
> 		p.keyBuf = p.curRawKey
> ```

Same change as before.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, @jbowens, and @petermattis)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 438 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIKTOhu0eK8XA0slgXm:-MIKTOhu0eK8XA0slgXn:b-ffu2kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0554209d416787e91d197c3b116a3fa6046e14b/pkg/storage/pebble_mvcc_scanner.go#L438)):*
> ```Go
> 			// about to advance. If this proves to be a problem later, we can extend
> 			// addAndAdvance to take an MVCCKey explicitly.
> 			p.curKey.Timestamp = metaTS
> ```

So this is where the test failure is happening. We mutate curKey without also mutating curRawKey.

There are two solutions here:

1) Encode `p.curKey` after this change into `p.keyBuf`, have `p.curRawKey` point to that, and then `addAndAdvance` would do the right thing. This is a bit of a hack, but the existing code is also a bit of a hack as the comment describes.

2) The cleaner way (and the one I'd prefer) is to add a `rawKey []byte` parameter to `addAndAdvance`. Here, you'd still encode the new `MVCCKey{p.curKey, metaTS}` into `p.keyBuf`, and then pass `p.keyBuf` into `addAndAdvance` as the raw key parameter. In all other calls of `addAndAdvance`, you can just pass `p.curRawKey`.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
Most of the changes here are a revert of https://reviewable.io/reviews/cockroachdb/cockroach/42210. I commented on some of the differences below.

I also ran the `MVCCScan_Pebble` benchmark (on a gceworker) comparing this branch to the current master. The results show a slight improvement in most tests but there are some tests showing some regression. Any thoughts/ideas?

```
name                                                    old time/op    new time/op    delta
MVCCScan_Pebble/rows=1/versions=1/valueSize=8-24          3.54µs ± 1%    3.48µs ± 1%  -1.69%  (p=0.000 n=9+10)
MVCCScan_Pebble/rows=1/versions=1/valueSize=64-24         3.97µs ± 2%    3.89µs ± 1%  -1.92%  (p=0.001 n=10+10)
MVCCScan_Pebble/rows=1/versions=1/valueSize=512-24        5.32µs ± 1%    5.20µs ± 2%  -2.19%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=1/versions=2/valueSize=8-24          3.80µs ± 1%    3.79µs ± 1%    ~     (p=0.088 n=9+9)
MVCCScan_Pebble/rows=1/versions=2/valueSize=64-24         4.30µs ± 2%    4.22µs ± 1%  -1.94%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=2/valueSize=512-24        8.56µs ± 1%    8.40µs ± 1%  -1.84%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=8-24         5.28µs ± 1%    5.25µs ± 1%    ~     (p=0.056 n=9+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=64-24        5.90µs ± 1%    5.92µs ± 2%    ~     (p=0.542 n=10+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=512-24       10.5µs ± 1%    10.4µs ± 1%  -0.90%  (p=0.029 n=9+9)
MVCCScan_Pebble/rows=1/versions=100/valueSize=8-24        14.4µs ± 1%    14.3µs ± 1%    ~     (p=0.183 n=10+10)
MVCCScan_Pebble/rows=1/versions=100/valueSize=64-24       14.4µs ± 1%    14.5µs ± 1%  +0.87%  (p=0.013 n=9+10)
MVCCScan_Pebble/rows=1/versions=100/valueSize=512-24      44.5µs ± 1%    45.4µs ± 1%  +2.19%  (p=0.000 n=8+8)
MVCCScan_Pebble/rows=10/versions=1/valueSize=8-24         7.10µs ± 1%    7.00µs ± 0%  -1.40%  (p=0.000 n=10+8)
MVCCScan_Pebble/rows=10/versions=1/valueSize=64-24        8.58µs ± 1%    8.51µs ± 1%  -0.84%  (p=0.019 n=10+10)
MVCCScan_Pebble/rows=10/versions=1/valueSize=512-24       12.0µs ± 1%    11.9µs ± 1%  -1.27%  (p=0.001 n=10+8)
MVCCScan_Pebble/rows=10/versions=2/valueSize=8-24         8.37µs ± 0%    8.28µs ± 1%  -1.01%  (p=0.000 n=8+10)
MVCCScan_Pebble/rows=10/versions=2/valueSize=64-24        10.0µs ± 1%     9.9µs ± 1%  -1.49%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=10/versions=2/valueSize=512-24       18.1µs ± 1%    17.6µs ± 1%  -2.33%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=10/versions=10/valueSize=8-24        17.1µs ± 2%    17.0µs ± 1%    ~     (p=0.971 n=10+10)
MVCCScan_Pebble/rows=10/versions=10/valueSize=64-24       19.2µs ± 1%    19.2µs ± 1%    ~     (p=0.971 n=10+10)
MVCCScan_Pebble/rows=10/versions=10/valueSize=512-24      32.7µs ± 0%    32.7µs ± 1%    ~     (p=0.829 n=8+10)
MVCCScan_Pebble/rows=10/versions=100/valueSize=8-24       57.9µs ± 1%    58.3µs ± 0%  +0.63%  (p=0.013 n=10+9)
MVCCScan_Pebble/rows=10/versions=100/valueSize=64-24      60.1µs ± 0%    60.5µs ± 0%  +0.79%  (p=0.000 n=9+9)
MVCCScan_Pebble/rows=10/versions=100/valueSize=512-24      186µs ± 2%     190µs ± 3%  +1.84%  (p=0.002 n=10+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=8-24        33.5µs ± 1%    33.1µs ± 1%  -1.25%  (p=0.001 n=9+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=64-24       40.0µs ± 1%    39.6µs ± 2%  -0.96%  (p=0.015 n=10+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=512-24      59.8µs ± 2%    59.5µs ± 1%    ~     (p=1.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=8-24        42.8µs ± 0%    42.6µs ± 1%  -0.54%  (p=0.003 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=64-24       50.1µs ± 1%    50.3µs ± 2%    ~     (p=0.631 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=512-24      87.9µs ± 3%    85.5µs ± 1%  -2.80%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=8-24        109µs ± 1%     110µs ± 1%  +0.90%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=64-24       119µs ± 1%     120µs ± 1%  +0.88%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=100/versions=10/valueSize=512-24      213µs ± 0%     212µs ± 0%  -0.36%  (p=0.011 n=8+9)
MVCCScan_Pebble/rows=100/versions=100/valueSize=8-24       454µs ± 1%     455µs ± 1%    ~     (p=0.631 n=10+10)
MVCCScan_Pebble/rows=100/versions=100/valueSize=64-24      465µs ± 1%     471µs ± 2%  +1.20%  (p=0.002 n=10+10)
MVCCScan_Pebble/rows=100/versions=100/valueSize=512-24    1.47ms ± 1%    1.48ms ± 1%  +0.97%  (p=0.014 n=7+6)

name                                                    old speed      new speed      delta
MVCCScan_Pebble/rows=1/versions=1/valueSize=8-24        2.26MB/s ± 0%  2.30MB/s ± 1%  +1.64%  (p=0.000 n=6+10)
MVCCScan_Pebble/rows=1/versions=1/valueSize=64-24       16.1MB/s ± 2%  16.4MB/s ± 1%  +1.97%  (p=0.001 n=10+10)
MVCCScan_Pebble/rows=1/versions=1/valueSize=512-24      96.3MB/s ± 1%  98.5MB/s ± 2%  +2.25%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=1/versions=2/valueSize=8-24        2.11MB/s ± 2%  2.11MB/s ± 1%    ~     (p=0.236 n=10+9)
MVCCScan_Pebble/rows=1/versions=2/valueSize=64-24       14.9MB/s ± 2%  15.2MB/s ± 1%  +1.97%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=2/valueSize=512-24      59.8MB/s ± 1%  60.9MB/s ± 1%  +1.87%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=8-24       1.51MB/s ± 1%  1.52MB/s ± 1%  +0.63%  (p=0.014 n=9+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=64-24      10.8MB/s ± 1%  10.8MB/s ± 2%    ~     (p=0.539 n=10+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=512-24     48.7MB/s ± 1%  49.1MB/s ± 1%  +0.92%  (p=0.034 n=9+9)
MVCCScan_Pebble/rows=1/versions=100/valueSize=8-24       557kB/s ± 1%   560kB/s ± 0%    ~     (p=0.353 n=10+7)
MVCCScan_Pebble/rows=1/versions=100/valueSize=64-24     4.43MB/s ± 2%  4.40MB/s ± 1%    ~     (p=0.057 n=10+10)
MVCCScan_Pebble/rows=1/versions=100/valueSize=512-24    11.5MB/s ± 1%  11.3MB/s ± 1%  -2.16%  (p=0.000 n=8+8)
MVCCScan_Pebble/rows=10/versions=1/valueSize=8-24       11.3MB/s ± 1%  11.4MB/s ± 0%  +1.42%  (p=0.000 n=10+8)
MVCCScan_Pebble/rows=10/versions=1/valueSize=64-24      74.6MB/s ± 1%  75.2MB/s ± 1%  +0.85%  (p=0.017 n=10+10)
MVCCScan_Pebble/rows=10/versions=1/valueSize=512-24      426MB/s ± 1%   431MB/s ± 1%  +1.29%  (p=0.001 n=10+8)
MVCCScan_Pebble/rows=10/versions=2/valueSize=8-24       9.57MB/s ± 0%  9.66MB/s ± 1%  +1.01%  (p=0.000 n=8+10)
MVCCScan_Pebble/rows=10/versions=2/valueSize=64-24      63.8MB/s ± 1%  64.8MB/s ± 1%  +1.51%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=10/versions=2/valueSize=512-24      283MB/s ± 1%   290MB/s ± 1%  +2.38%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=10/versions=10/valueSize=8-24      4.69MB/s ± 2%  4.69MB/s ± 1%    ~     (p=0.870 n=10+10)
MVCCScan_Pebble/rows=10/versions=10/valueSize=64-24     33.3MB/s ± 1%  33.3MB/s ± 1%    ~     (p=0.985 n=10+10)
MVCCScan_Pebble/rows=10/versions=10/valueSize=512-24     157MB/s ± 0%   157MB/s ± 1%    ~     (p=0.829 n=8+10)
MVCCScan_Pebble/rows=10/versions=100/valueSize=8-24     1.38MB/s ± 1%  1.37MB/s ± 1%  -0.72%  (p=0.015 n=10+10)
MVCCScan_Pebble/rows=10/versions=100/valueSize=64-24    10.7MB/s ± 0%  10.6MB/s ± 0%  -0.78%  (p=0.000 n=9+9)
MVCCScan_Pebble/rows=10/versions=100/valueSize=512-24   27.5MB/s ± 2%  27.0MB/s ± 3%  -1.80%  (p=0.002 n=10+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=8-24      23.9MB/s ± 1%  24.2MB/s ± 1%  +1.28%  (p=0.000 n=9+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=64-24      160MB/s ± 1%   162MB/s ± 2%  +0.97%  (p=0.015 n=10+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=512-24     857MB/s ± 2%   860MB/s ± 1%    ~     (p=1.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=8-24      18.7MB/s ± 0%  18.8MB/s ± 1%  +0.56%  (p=0.003 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=64-24      128MB/s ± 1%   127MB/s ± 2%    ~     (p=0.631 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=512-24     583MB/s ± 3%   599MB/s ± 1%  +2.84%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=8-24     7.34MB/s ± 1%  7.27MB/s ± 1%  -0.93%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=64-24    53.8MB/s ± 1%  53.3MB/s ± 1%  -0.87%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=100/versions=10/valueSize=512-24    241MB/s ± 0%   241MB/s ± 0%  +0.36%  (p=0.011 n=8+9)
MVCCScan_Pebble/rows=100/versions=100/valueSize=8-24    1.77MB/s ± 1%  1.76MB/s ± 1%    ~     (p=0.287 n=9+10)
MVCCScan_Pebble/rows=100/versions=100/valueSize=64-24   13.8MB/s ± 1%  13.6MB/s ± 2%  -1.19%  (p=0.003 n=10+10)
MVCCScan_Pebble/rows=100/versions=100/valueSize=512-24  34.9MB/s ± 1%  34.6MB/s ± 1%  -0.96%  (p=0.013 n=7+6)
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-:-MIPwLs53jkvhr0DaRu8:b-4adnbv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 735 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI0W5J75TQTnoiyt2Vm:-MIQ7yxPCQc9aq_NcceK:b-g8acnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b6dabed3269c933c66bba722f80032faced0b1/pkg/storage/pebble_mvcc_scanner.go#L735)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Thanks, will try to compare the results once I can get these tests to pass. 
</blockquote></details>

Made the change and ran a quick benchmark:
```
name                                                  old time/op    new time/op    delta
MVCCScan_Pebble/rows=1/versions=2/valueSize=8-24        3.76µs ± 1%    3.78µs ± 1%    ~     (p=0.075 n=9+10)
MVCCScan_Pebble/rows=10/versions=2/valueSize=8-24       8.32µs ± 1%    8.27µs ± 1%  -0.60%  (p=0.016 n=10+9)
MVCCScan_Pebble/rows=100/versions=2/valueSize=8-24      42.8µs ± 1%    42.7µs ± 1%    ~     (p=0.985 n=10+9)
MVCCScan_Pebble/rows=1000/versions=2/valueSize=8-24      366µs ± 0%     366µs ± 0%    ~     (p=0.604 n=9+10)
MVCCScan_Pebble/rows=10000/versions=2/valueSize=8-24    3.34ms ± 0%    3.35ms ± 0%    ~     (p=0.515 n=8+10)

name                                                  old speed      new speed      delta
MVCCScan_Pebble/rows=1/versions=2/valueSize=8-24      2.13MB/s ± 1%  2.12MB/s ± 1%    ~     (p=0.064 n=9+10)
MVCCScan_Pebble/rows=10/versions=2/valueSize=8-24     9.62MB/s ± 1%  9.67MB/s ± 0%  +0.58%  (p=0.020 n=10+9)
MVCCScan_Pebble/rows=100/versions=2/valueSize=8-24    18.7MB/s ± 1%  18.7MB/s ± 1%    ~     (p=1.000 n=10+9)
MVCCScan_Pebble/rows=1000/versions=2/valueSize=8-24   21.9MB/s ± 0%  21.9MB/s ± 0%    ~     (p=0.704 n=9+10)
MVCCScan_Pebble/rows=10000/versions=2/valueSize=8-24  23.9MB/s ± 1%  23.9MB/s ± 0%    ~     (p=0.344 n=9+10)
```
It seems to be slightly better (but could be negligible). It does avoid calling `enginepb.SplitMVCCKey(p.curRawKey)` and splitting other parts so decided to go ahead with the change.

The previous change called `enginepb.SplitMVCCKey(p.curRawKey)` which is unnecessary since we only need `p.curKey.Key` (and not the TS)
___
*[pkg/storage/pebble_mvcc_scanner.go, line 407 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI4eicC645RTSvoPfGc:-MIQD_EwE4yCy6E-2CIc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/pebble_mvcc_scanner.go#L407)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

You can skip this encoding and reuse `p.curRawKey` here.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 385 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIK0C1Y7p9o0dWKkNtv:-MIQDYX2DDtYI4ffN9jk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3ab85c89003ea9ec13b0664e8b0c8d1c126cbb/pkg/storage/pebble_mvcc_scanner.go#L385)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

`p.keyBuf` is meant to always point to a clean buffer owned by `pebbleMVCCScanner` and not by the iterator. So you'd want to remove this assignment and use `p.curRawKey` directly below.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 406 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIK0my_BUj2imaxN-7M:-MIQDZcq8_4CwwDUvPds:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3ab85c89003ea9ec13b0664e8b0c8d1c126cbb/pkg/storage/pebble_mvcc_scanner.go#L406)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Same change as before.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 438 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIKTOhu0eK8XA0slgXm:-MIQDamH65JAnDkY6WJZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0554209d416787e91d197c3b116a3fa6046e14b/pkg/storage/pebble_mvcc_scanner.go#L438)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

So this is where the test failure is happening. We mutate curKey without also mutating curRawKey.

There are two solutions here:

1) Encode `p.curKey` after this change into `p.keyBuf`, have `p.curRawKey` point to that, and then `addAndAdvance` would do the right thing. This is a bit of a hack, but the existing code is also a bit of a hack as the comment describes.

2) The cleaner way (and the one I'd prefer) is to add a `rawKey []byte` parameter to `addAndAdvance`. Here, you'd still encode the new `MVCCKey{p.curKey, metaTS}` into `p.keyBuf`, and then pass `p.keyBuf` into `addAndAdvance` as the raw key parameter. In all other calls of `addAndAdvance`, you can just pass `p.curRawKey`.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 649 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIQEDNI9zQl5GplW3G3:-MIQEDNI9zQl5GplW3G4:b1mn7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8b5ff5eb03d0622beb000466b67d6a1cbb4cfb1/pkg/storage/pebble_mvcc_scanner.go#L649)):*
> ```Go
> 	}
> 
> 	p.curRawKey = p.parent.UnsafeRawKey()
> ```

This change is different from previous due the change in the iterator interface. `p.parent.UnsafeKey()` can be called instead of decoding the RawKey.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
Looks great so far! This is almost ready.

The performance improvements are in line with what I expected at small row counts - not sure why the large value sizes / large row counts regressed a bit, but I have one suggestion that could probably fix that. Feel free to make that change and re-run a small number of the slower benchmarks to ensure it's fixed now.

> Reviewed 1 of 6 files at r2, 3 of 5 files at r3, 1 of 2 files at r4, 1 of 2 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-:-MIQlp_u8gfBnf3S6r4Q:bu3etu3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @jbowens, and @petermattis)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 737 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI4eZukBa2nlzyxnnaJ:-MIQl9408RfGT5ZpY0Vr:bwx98mn) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/pebble_mvcc_scanner.go#L737)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Made the change, but tests were still failing. I am trying to debug but haven't had any luck yet.
</blockquote></details>

Looks good to me! You can expand the comment right before this line, so that it directly name-drops SplitMVCCKey if someone searches that up:

```
// The raw key is always a prefix of the encoded MVCC key. Take advantage of this to
// sub-slice the raw key directly, instead of calling SplitMVCCKey.
```
___
*[pkg/storage/pebble_mvcc_scanner.go, line 582 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIQntcr5VE3a-gPzrBL:-MIQntcr5VE3a-gPzrBM:b-25cgwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8b5ff5eb03d0622beb000466b67d6a1cbb4cfb1/pkg/storage/pebble_mvcc_scanner.go#L582)):*
> ```Go
> // p.tombstones is true. Advances to the next key unless we've reached the max
> // results limit.
> func (p *pebbleMVCCScanner) addAndAdvance(val []byte, rawKey []byte) bool {
> ```

I would move the `rawKey` to be before the `val`. That seems more natural; key before value.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 650 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIQoQq802h_qEh6oTXb:-MIQoQq802h_qEh6oTXc:bpjkpxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8b5ff5eb03d0622beb000466b67d6a1cbb4cfb1/pkg/storage/pebble_mvcc_scanner.go#L650)):*
> ```Go
> 
> 	p.curRawKey = p.parent.UnsafeRawKey()
> 	p.curKey = p.parent.UnsafeKey()
> ```

I wonder if the extra overhead of calling both UnsafeRawKey and UnsafeKey is causing the slight regression. UnsafeKey does an extra validity check, which is useless since you've already done it above. If you replace this with `p.curKey, err = DecodeMVCCKey(p.curRawKey)` and then panic on the error, maybe you'd get slightly better performance in the slower cases.
___
*[pkg/storage/rocksdb.go, line 2035 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIQlOpAARfvHhcZ7rnj:-MIQlOpAARfvHhcZ7rnk:b-98lhx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8b5ff5eb03d0622beb000466b67d6a1cbb4cfb1/pkg/storage/rocksdb.go#L2035)):*
> ```Go
> 
> func (r *rocksDBIterator) UnsafeRawKey() []byte {
> 	panic("unimplemented")
> ```

Nit: You could change this to "unimplemented; only meant to be used with Pebble"


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
Re-ran benchmarks using `DecodeMVCCKey(p.curRawKey)` in `updateCurrent()`, results show a much more consistent improvement than before:

```
name                                                    old time/op    new time/op     delta
MVCCScan_Pebble/rows=1/versions=1/valueSize=8-24          3.60µs ± 1%     3.44µs ± 1%  -4.45%  (p=0.000 n=9+10)
MVCCScan_Pebble/rows=1/versions=1/valueSize=64-24         3.99µs ± 1%     3.79µs ± 1%  -5.10%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=1/valueSize=512-24        5.28µs ± 2%     5.07µs ± 2%  -4.03%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=2/valueSize=8-24          3.84µs ± 1%     3.70µs ± 1%  -3.60%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=2/valueSize=64-24         4.29µs ± 1%     4.09µs ± 1%  -4.78%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=2/valueSize=512-24        8.51µs ± 1%     7.96µs ± 1%  -6.47%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=8-24         5.29µs ± 2%     5.06µs ± 1%  -4.38%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=64-24        5.92µs ± 1%     5.56µs ± 1%  -6.00%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=1/versions=10/valueSize=512-24       10.0µs ± 1%      9.8µs ± 2%  -1.86%  (p=0.008 n=9+9)
MVCCScan_Pebble/rows=1/versions=100/valueSize=8-24        14.2µs ± 1%     13.6µs ± 4%  -4.19%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=100/valueSize=64-24       14.3µs ± 1%     13.8µs ± 1%  -3.63%  (p=0.000 n=9+10)
MVCCScan_Pebble/rows=1/versions=100/valueSize=512-24      44.4µs ± 2%     43.4µs ± 4%  -2.38%  (p=0.006 n=10+8)
MVCCScan_Pebble/rows=10/versions=1/valueSize=8-24         7.15µs ± 0%     6.81µs ± 1%  -4.69%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=10/versions=1/valueSize=64-24        8.61µs ± 1%     8.15µs ± 0%  -5.28%  (p=0.000 n=10+8)
MVCCScan_Pebble/rows=10/versions=1/valueSize=512-24       12.0µs ± 1%     11.4µs ± 1%  -4.63%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=10/versions=2/valueSize=8-24         8.35µs ± 1%     8.07µs ± 1%  -3.39%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=10/versions=2/valueSize=64-24        10.0µs ± 2%      9.6µs ± 1%  -3.76%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=10/versions=2/valueSize=512-24       17.8µs ± 2%     17.3µs ± 1%  -2.79%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=10/versions=10/valueSize=8-24        16.9µs ± 2%     16.6µs ± 0%  -1.93%  (p=0.000 n=10+8)
MVCCScan_Pebble/rows=10/versions=10/valueSize=64-24       18.9µs ± 1%     18.7µs ± 1%  -1.27%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=10/versions=10/valueSize=512-24      31.3µs ± 2%     31.5µs ± 1%  +0.73%  (p=0.037 n=10+9)
MVCCScan_Pebble/rows=10/versions=100/valueSize=8-24       58.3µs ± 1%     58.1µs ± 1%    ~     (p=0.075 n=10+10)
MVCCScan_Pebble/rows=10/versions=100/valueSize=64-24      60.6µs ± 1%     60.7µs ± 1%    ~     (p=1.000 n=10+10)
MVCCScan_Pebble/rows=10/versions=100/valueSize=512-24      187µs ± 1%      185µs ± 1%  -0.94%  (p=0.001 n=9+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=8-24        33.9µs ± 1%     32.8µs ± 1%  -3.16%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=64-24       41.1µs ± 3%     39.4µs ± 1%  -3.99%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=512-24      60.0µs ± 2%     59.3µs ± 1%  -1.24%  (p=0.009 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=8-24        43.4µs ± 1%     41.8µs ± 0%  -3.80%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=100/versions=2/valueSize=64-24       50.6µs ± 1%     49.1µs ± 1%  -2.97%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=512-24      86.2µs ± 1%     85.2µs ± 1%  -1.07%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=8-24        111µs ± 2%      107µs ± 1%  -3.67%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=64-24       119µs ± 1%      116µs ± 0%  -2.10%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=512-24      203µs ± 1%      210µs ± 2%  +3.23%  (p=0.000 n=9+10)
MVCCScan_Pebble/rows=100/versions=100/valueSize=8-24       459µs ± 1%      457µs ± 0%    ~     (p=0.079 n=10+9)
MVCCScan_Pebble/rows=100/versions=100/valueSize=64-24      468µs ± 0%      470µs ± 1%  +0.40%  (p=0.043 n=10+10)
MVCCScan_Pebble/rows=100/versions=100/valueSize=512-24    1.46ms ± 3%     1.44ms ± 3%    ~     (p=0.063 n=10+10)
MVCCScan_Pebble/rows=1000/versions=1/valueSize=8-24        287µs ± 1%      276µs ± 0%  -3.99%  (p=0.000 n=10+7)

name                                                    old speed      new speed       delta
MVCCScan_Pebble/rows=1/versions=1/valueSize=8-24        2.22MB/s ± 1%   2.33MB/s ± 1%  +4.73%  (p=0.000 n=9+10)
MVCCScan_Pebble/rows=1/versions=1/valueSize=64-24       16.0MB/s ± 1%   16.9MB/s ± 1%  +5.38%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=1/valueSize=512-24      96.9MB/s ± 2%  101.0MB/s ± 2%  +4.21%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=2/valueSize=8-24        2.09MB/s ± 1%   2.16MB/s ± 1%  +3.61%  (p=0.000 n=9+10)
MVCCScan_Pebble/rows=1/versions=2/valueSize=64-24       14.9MB/s ± 1%   15.7MB/s ± 1%  +5.02%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=2/valueSize=512-24      60.2MB/s ± 1%   64.4MB/s ± 1%  +6.92%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=8-24       1.51MB/s ± 2%   1.58MB/s ± 1%  +4.56%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=64-24      10.8MB/s ± 1%   11.5MB/s ± 1%  +6.41%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=1/versions=10/valueSize=512-24     51.5MB/s ± 1%   52.4MB/s ± 2%  +1.92%  (p=0.008 n=9+9)
MVCCScan_Pebble/rows=1/versions=100/valueSize=8-24       563kB/s ± 1%    589kB/s ± 5%  +4.62%  (p=0.001 n=10+10)
MVCCScan_Pebble/rows=1/versions=100/valueSize=64-24     4.46MB/s ± 1%   4.63MB/s ± 1%  +3.78%  (p=0.000 n=9+10)
MVCCScan_Pebble/rows=1/versions=100/valueSize=512-24    11.5MB/s ± 2%   11.8MB/s ± 4%  +2.47%  (p=0.006 n=10+8)
MVCCScan_Pebble/rows=10/versions=1/valueSize=8-24       11.2MB/s ± 1%   11.7MB/s ± 1%  +4.93%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=10/versions=1/valueSize=64-24      74.4MB/s ± 1%   78.5MB/s ± 0%  +5.56%  (p=0.000 n=10+8)
MVCCScan_Pebble/rows=10/versions=1/valueSize=512-24      427MB/s ± 1%    447MB/s ± 1%  +4.86%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=10/versions=2/valueSize=8-24       9.58MB/s ± 1%   9.92MB/s ± 1%  +3.52%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=10/versions=2/valueSize=64-24      64.3MB/s ± 2%   66.8MB/s ± 1%  +3.90%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=10/versions=2/valueSize=512-24      288MB/s ± 2%    296MB/s ± 1%  +2.86%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=10/versions=10/valueSize=8-24      4.73MB/s ± 2%   4.83MB/s ± 1%  +1.97%  (p=0.000 n=10+8)
MVCCScan_Pebble/rows=10/versions=10/valueSize=64-24     33.9MB/s ± 1%   34.3MB/s ± 1%  +1.29%  (p=0.000 n=10+9)
MVCCScan_Pebble/rows=10/versions=10/valueSize=512-24     164MB/s ± 2%    162MB/s ± 1%  -0.73%  (p=0.036 n=10+9)
MVCCScan_Pebble/rows=10/versions=100/valueSize=8-24     1.37MB/s ± 1%   1.38MB/s ± 1%    ~     (p=0.270 n=10+10)
MVCCScan_Pebble/rows=10/versions=100/valueSize=64-24    10.6MB/s ± 1%   10.6MB/s ± 1%    ~     (p=0.927 n=10+10)
MVCCScan_Pebble/rows=10/versions=100/valueSize=512-24   27.4MB/s ± 1%   27.7MB/s ± 1%  +0.95%  (p=0.001 n=9+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=8-24      23.6MB/s ± 1%   24.4MB/s ± 1%  +3.26%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=64-24      156MB/s ± 3%    162MB/s ± 1%  +4.14%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=512-24     853MB/s ± 2%    864MB/s ± 1%  +1.25%  (p=0.009 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=8-24      18.4MB/s ± 1%   19.2MB/s ± 1%  +4.05%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=64-24      126MB/s ± 1%    130MB/s ± 1%  +3.06%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=2/valueSize=512-24     594MB/s ± 1%    601MB/s ± 1%  +1.08%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=8-24     7.21MB/s ± 2%   7.48MB/s ± 1%  +3.77%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=64-24    54.0MB/s ± 1%   55.1MB/s ± 0%  +2.15%  (p=0.000 n=10+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=512-24    252MB/s ± 1%    244MB/s ± 2%  -3.12%  (p=0.000 n=9+10)
MVCCScan_Pebble/rows=100/versions=100/valueSize=8-24    1.75MB/s ± 1%   1.75MB/s ± 0%    ~     (p=0.092 n=10+8)
MVCCScan_Pebble/rows=100/versions=100/valueSize=64-24   13.7MB/s ± 1%   13.6MB/s ± 1%  -0.39%  (p=0.041 n=10+10)
MVCCScan_Pebble/rows=100/versions=100/valueSize=512-24  35.2MB/s ± 3%   35.5MB/s ± 3%    ~     (p=0.066 n=10+10)
MVCCScan_Pebble/rows=1000/versions=1/valueSize=8-24     27.8MB/s ± 1%   29.0MB/s ± 0%  +4.15%  (p=0.000 n=10+7)
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-:-MIUze5q3whhv5hU9rqX:b-q727od)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 737 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MI4eZukBa2nlzyxnnaJ:-MIUzL-jCuS5Ef-6QeA8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b181b5e4ce4bb33ae0e2fa8549c42a81bdb5b1/pkg/storage/pebble_mvcc_scanner.go#L737)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Looks good to me! You can expand the comment right before this line, so that it directly name-drops SplitMVCCKey if someone searches that up:

```
// The raw key is always a prefix of the encoded MVCC key. Take advantage of this to
// sub-slice the raw key directly, instead of calling SplitMVCCKey.
```
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 582 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIQntcr5VE3a-gPzrBL:-MIUzVhK081kR4O9jAE5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8b5ff5eb03d0622beb000466b67d6a1cbb4cfb1/pkg/storage/pebble_mvcc_scanner.go#L582)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I would move the `rawKey` to be before the `val`. That seems more natural; key before value.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 650 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIQoQq802h_qEh6oTXb:-MIUzT7LBGdl_YOQvKXW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8b5ff5eb03d0622beb000466b67d6a1cbb4cfb1/pkg/storage/pebble_mvcc_scanner.go#L650)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I wonder if the extra overhead of calling both UnsafeRawKey and UnsafeKey is causing the slight regression. UnsafeKey does an extra validity check, which is useless since you've already done it above. If you replace this with `p.curKey, err = DecodeMVCCKey(p.curRawKey)` and then panic on the error, maybe you'd get slightly better performance in the slower cases.
</blockquote></details>

Done.
___
*[pkg/storage/rocksdb.go, line 2035 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIQlOpAARfvHhcZ7rnj:-MIV0NQ3DfmeNLHM50NB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8b5ff5eb03d0622beb000466b67d6a1cbb4cfb1/pkg/storage/rocksdb.go#L2035)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Nit: You could change this to "unimplemented; only meant to be used with Pebble"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> great work!

The only remaining suggestions are really specific nits, feel free to fix it and then merge after that without waiting for another LGTM from me. Though Peter may also want to take a look in the meantime, so that gives him some time to do that too.

> Reviewed 2 of 2 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-:-MIV5nrGCeeh5OA-pSLM:br9zjsi)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, @jbowens, and @petermattis)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 579 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIV5hJi0Dq9cSIk8q0U:-MIV5hJi0Dq9cSIk8q0V:bxym4l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/12d1c9202205a747eca1d3f2c67aebdb100201e8/pkg/storage/pebble_mvcc_scanner.go#L579)):*
> ```Go
> }
> 
> // Adds the specified value to the result set, excluding tombstones unless
> ```

Nit: Could be changed to `Adds the specified key and value to ...`
___
*[pkg/storage/pebble_mvcc_scanner.go, line 651 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIV5PDj9WHUAUAg64jX:-MIV5PDj9WHUAUAg64jY:bs35hf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/12d1c9202205a747eca1d3f2c67aebdb100201e8/pkg/storage/pebble_mvcc_scanner.go#L651)):*
> ```Go
> 	p.curRawKey = p.parent.UnsafeRawKey()
> 
> 	mvccKey, err := DecodeMVCCKey(p.curRawKey)
> ```

Nittiest of nits: You could just do:

```
var err error
p.curKey, err = ...
if err != nil {
   panic(err)
}
```

Wouldn't be surprised if it's just a hair faster :joy:


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
TYFR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-:-MIVpsaHCyZuxR7vLRVo:b-80slv9)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @jbowens, and @petermattis)



<!-- Sent from Reviewable.io -->

aadityasondhi commented:
bors r+

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi, @itsbilal, and @jbowens)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 652 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIVreg4CMjlrddGgWdU:-MIVreg4CMjlrddGgWdV:bqj265d) ([raw file](https://github.com/cockroachdb/cockroach/blob/023663819426e74873807ff7c541cba9739e434b/pkg/storage/pebble_mvcc_scanner.go#L652)):*
> ```Go
> 
> 	var err error
> 	p.curKey, err = DecodeMVCCKey(p.curRawKey)
> ```

Go doesn't currently have mid-stack inlining (at least, I don't think that has change). There might be a small perf advantage to writing this as:

```
p.curKey.Key, p.curKey.Timestamp, err = enginepb.DecodeKey(p.curRawKey)
```

I'll note that is how this was written prior to https://reviewable.io/reviews/cockroachdb/cockroach/42210.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi, @itsbilal, and @jbowens)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 652 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIVreg4CMjlrddGgWdU:-MIZ0yWsEGPtciajHGXa:b-rgfzkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/023663819426e74873807ff7c541cba9739e434b/pkg/storage/pebble_mvcc_scanner.go#L652)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Go doesn't currently have mid-stack inlining (at least, I don't think that has change). There might be a small perf advantage to writing this as:

```
p.curKey.Key, p.curKey.Timestamp, err = enginepb.DecodeKey(p.curRawKey)
```

I'll note that is how this was written prior to https://reviewable.io/reviews/cockroachdb/cockroach/42210.
</blockquote></details>

@aadityasondhi and @itsbilal I want to make sure this comment is not forgotten. There might be a tiny bit of performance to be had.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi, @itsbilal, and @jbowens)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 652 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIVreg4CMjlrddGgWdU:-MIZT0Vz94isfXbYlk9_:b-y9j1mv) ([raw file](https://github.com/cockroachdb/cockroach/blob/023663819426e74873807ff7c541cba9739e434b/pkg/storage/pebble_mvcc_scanner.go#L652)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

@aadityasondhi and @itsbilal I want to make sure this comment is not forgotten. There might be a tiny bit of performance to be had.
</blockquote></details>

https://dave.cheney.net/2020/05/02/mid-stack-inlining-in-go suggests it does have mid-stack inlining. Am I missing something?


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @jbowens)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 652 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIVreg4CMjlrddGgWdU:-MIZY4Gv-F1lcnIVPvRA:bwz7enu) ([raw file](https://github.com/cockroachdb/cockroach/blob/023663819426e74873807ff7c541cba9739e434b/pkg/storage/pebble_mvcc_scanner.go#L652)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

https://dave.cheney.net/2020/05/02/mid-stack-inlining-in-go suggests it does have mid-stack inlining. Am I missing something?
</blockquote></details>

Thanks for pointing that out. I can run a quick benchmark to see if calling `enginepb.DecodeKey(p.curRawKey)` helps performance but it does seem like the compiler would automatically do the mid-stack inlining


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @jbowens)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 652 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIVreg4CMjlrddGgWdU:-MIZVhh73Y6Wp-jza-XU:b-8yjeop) ([raw file](https://github.com/cockroachdb/cockroach/blob/023663819426e74873807ff7c541cba9739e434b/pkg/storage/pebble_mvcc_scanner.go#L652)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Thanks for pointing that out. I can run a quick benchmark to see if calling `enginepb.DecodeKey(p.curRawKey)` helps performance but it does seem like the compiler would automatically do the mid-stack inlining
</blockquote></details>

I thought they added mid-stack inlining and then rolled it back. I could be missing something too. You can look at the compiled code to see if it is inlining or not. I'd only make this change if it is beneficial.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54699)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @jbowens)

___
*[pkg/storage/pebble_mvcc_scanner.go, line 652 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54699#-MIVreg4CMjlrddGgWdU:-MIZplJD-mrouk09DQvo:b-1n6jjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/023663819426e74873807ff7c541cba9739e434b/pkg/storage/pebble_mvcc_scanner.go#L652)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I thought they added mid-stack inlining and then rolled it back. I could be missing something too. You can look at the compiled code to see if it is inlining or not. I'd only make this change if it is beneficial.
</blockquote></details>

I have verified that the compiled code is not calling the `DecodeMVCCKey` function so I believe that the inlining is working. I am going to leave this call as-is based on this finding.


<!-- Sent from Reviewable.io -->
